sbin/i386/ft/ft.8
.\" Copyright (c) 1980, 1989, 1991 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ft.8
.\"	$Id: ft.8,v 1.2.6.2 1998/07/17 20:13:18 jkh Exp $
.\"
.Dd February 7, 1994
.Dt FT 8
.Os BSD 4
.Sh NAME
.Nm ft
.Nd QIC 40/80 floppy tape drive controller
.Sh SYNOPSIS
.Nm ft
.Op Fl f Ar tape
.Op Fl r
.Op Ar description
.Sh DESCRIPTION
The 
.Nm
command allows multi-volume dump, extract, and view of tape labels, for
any pre-formatted QIC-40/80 tapes.  It is totally system dependent,
and has nothing to do with the QIC standards.
.Pp
The following options are available:
.Bl -tag -width indent
.It Fl f Ar tape
Read or write the specified
.Ar tape
(default is
.Pa /dev/rft0
unless the environment variable
.Ev TAPE
is set).
.It Fl r
Do retension tape.
.El
.Pp
.Nm Ft
is used primarily as a filter for tape i/o.
For example, to save and compress the
.Pa /usr
directory to tape:
.Bd -literal -offset indent
% tar cvzf - /usr | ft "/usr save"
.Ed
.Pp
To extract /usr from tape:
.Bd -literal -offset indent
% ft | tar xvzf -
.Ed
.\" .Sh SEE ALSO
.\" .Xr qtar 1
.Sh BUGS
Formatting/Verifying is in the works.  You will need to use your
existing backup program to do this for the time being.
.Sh NOTES
The floppy tape driver supports tape drives such as the Colorado
Jumbo, Mountain Summit Express, some Archive/Conner models, and
probably many others.  These tape drives attach between your floppy
disk controller card and your existing floppy disks' ribbon cable.
This driver does not currently support attachments via a proprietary
tape controller card or by the parallel port.
.Pp
QIC-40/80 drives are more CPU intensive than a SCSI drive.  This is
really only a factor if your machine is networked or has multiple concurrent
users.  For personal use (i.e. your typical home Unix user), response time
is perfectly acceptable.  The tape drives cannot detect write errors.
Instead, they make up for it by using CRC's, error correction, and bad
spot mapping.  Formatting time is extremely long because of this.  The
drive makes a first pass over the entire tape writing out sectors.  It
then makes a second pass at a slower rate than usual (for sensitivity)
to detect bad spots on the tape.  Typically it takes an hour to format
a single QIC-80 (120Mb uncompressed) tape.
.Sh AUTHORS
.An Steve Gerakines Aq steve2@genesis.nred.ma.us


NEWS
# Major changes between releases

## Changes in version 0.15.0

**Released on 2026/10/07.**

### Bugfixes

* #265: Null cleanups
* #271: kyua: Try harder to delete directories.
* #300: Implement utils::config::positive_int_node::deep_copy

### Developer Workflow Changes

* #254: Add initial GitHub Actions support for kyua
* #288: [CI] remove legacy CI scripts and files
* #291: Remove Travis CI config
* #298: Add sanitizers and code coverage support
* #301: Add FreeBSD CI support via vmactions
* #305: Automate uploading release assets

### Feature Enhancements

* #203: Add GoogleTest support
* #230: Do not count skipped as passed in test cmd
* #260: Support require.diskspace ATF metadata
* #261: Pass unprivileged user config prop to ATF using all known names
* #266: Add "kyua debug -p" option
* #270: Add required_kmods metadata
* #272: Make "test" command actually report about workdir cleanup issues
* #273: Make "debug -p" skip writing stdout/stderr to tmp files
* #277: kyua-debug: Add -P option
* #290: Require C++-20 in order to build kyua
* #307: Upstream multiple items from FreeBSD

### Documentation

* #258: kyuafile.5: Add ATF metadata mapping reference
* #275: kyua manuals: Correct --results-file short form
* #279: docs: Standardize document description tone

### Miscellaneous

* #264: Switch from std::auto_ptr<> to std::unique_ptr<>
* #269: Fix cmd_about_test:topic__version test case
* #297: Reduce overall lutok dependency
* #312: Replace KYUA_ATTRIBUTE_{NORETURN,UNUSED}} with C++ attributes
* #318: utils::fs: remove umount(8) support

**Full Changelog**: https://github.com/freebsd/lutok/compare/kyua-0.14.1...kyua-0.15.0

## Changes in version 0.14.1

**Released on March 29th, 2025.**

* Issue #263: fix release tarball by relying on `make dist`.

* Document ATF to Kyua mappings for metadata properties.

* Support `require.diskspace` ATF metadata property.

* Support backwards compatible `unprivileged_user` metadata property for ease
  of use dealing with ATF vs Kyua.

* Fix building from source by tracking `utils/process/Kyuafile.in` and
  `engine/execenv/execenv_fwd.hpp` with automake.

## Changes in version 0.14

**Released on December 31st, 2024.**

* Replace `--with-atf` with `--enable-atf`. The latter must be specified
  to enable ATF support and the Kyua tests by proxy.

* Require newer versions of ATF (0.21) and Lutok (0.5). This is being done
  as part of a multi-phase deprecation of pre-C++11 environments.

* Vendor the ATF m4 files from the freebsd/atf repo.

* Translate `is_exclusive` metadata to its ATF counterpart.

* Issue #236: fix accidental test forkbomb in specific user environments
  with `integration` tests by avoiding infinite recursion calling kyua.

* Fix the build on Ubuntu 24.04.

- execenv support added. This FreeBSD-specific feature allows tests to be
  executed in jails.

* Issue #169: Remove deprecated `readdir_r` use.

* Fix build issue on more recent versions of MacOS.

* Fix build issue on IllumOS.

* Fix issues with PID reuse in testcases.

* Escape characters captured from stderr and stdout in HTML output.

* `integration/cmd_report_test:verbose`: fix regular expression used in
  tests which would break in certain user environments.

* Add Cirrus CI support for testing Kyua on FreeBSD.

* Remove `UTILS_UNUSED_PARAM` macro.

* Fix document rendering with newer versions of doxygen.

* Issue #152: explicitly require C++11 language features when compiling Kyua.
  Remove `std::shared_ptr` compatibility header, `utils/shared_ptr.hpp`.

* Fix `igor` and `mandoc -Tlint` reported issues with generated manpages.

* Issue #167: fix typos in kyua(1) manpage.

## Changes in version 0.13

**Released on August 26th, 2016.**

* Fixed execution of test cases as an unprivileged user, at least under
  NetBSD 7.0.  Kyua-level failures were probably a regression introduced
  in Kyua 0.12, but the underlying may have existed for much longer:
  test cases might have previously failed for mysterious reasons when
  running under an unprivileged user.

* Issue #134: Fixed metadata test broken on 32-bit platforms.

* Issue #139: Added per-test case start/end timestamps to all reports.

* Issue #156: Fixed crashes due to the invalid handling of cleanup
  routine data and triggered by the reuse of PIDs in long-running Kyua
  instances.

* Issue #159: Fixed TAP parser to ignore case while matching `TODO` and
  `SKIP` directives, and to also recognize `Skipped`.

* Fixed potential crash due to a race condition in the unprogramming of
  timers to control test deadlines.


## Changes in version 0.12

**Released on November 22nd, 2015.**

This is a huge release and marks a major milestone for Kyua as it finally
implements a long-standing feature request: the ability to execute test
cases in parallel.  This is a big deal because test cases are rarely
CPU-bound: running them in parallel yields much faster execution times for
large test suites, allowing faster iteration of changes during development.

As an example: the FreeBSD test suite as of this date contains 3285 test
cases.  With sequential execution, a full test suite run takes around 12
minutes to complete, whereas on a 4-core machine with a high level of
parallelism it takes a little over 1 minute.

Implementing parallel execution required rewriting most of Kyua's core and
partly explains explains why there has not been a new release for over a
year.  The current implementation is purely subprocess-based, which works
but has some limitations and has resulted in a core that is really complex
and difficult to understand.  Future versions will investigate the use of
threads instead for a simplified programming model and additional
parallelization possibilities.

* Issue #2: Implemented support to execute test cases in parallel when
  invoking `kyua test`.  Parallel execution is *only* enabled when the new
  `parallelism` configuration variable is set to a value greater than `1`.

sbin/i386/ft/ft.c
/*
 *  Copyright (c) 1993, 1994 Steve Gerakines
 *
 *  This is freely redistributable software.  You may do anything you
 *  wish with it, so long as the above notice stays intact.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS
 *  OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *  DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT,
 *  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 *  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 *  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 *  IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *
 *  ft.c - simple floppy tape filter
 *
 *  06/07/94 v1.0 ++sg
 *  Added support for tape retension.  Added retries for ecc failures.
 *  Moved to release.
 *
 *  01/28/94 v0.3b (Jim Babb)
 *  Fixed bug when all sectors in a segment are marked bad.
 *
 *  10/30/93 v0.3
 *  Minor revisions.  Seems pretty stable.
 *
 *  09/02/93 v0.2 pl01
 *  Initial revision.
 *
 *  usage: ft [ -f tape ] [ -r ] [ description ]
 */

#ifndef lint
static const char rcsid[] =
	"$Id: ft.c,v 1.3.6.1 1998/07/17 20:13:18 jkh Exp $";
#endif /* not lint */

#include <err.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sys/ftape.h>

#define	DEFQIC	"/dev/rft0"

char buff[QCV_SEGSIZE];			/* scratch buffer */
char hbuff[QCV_SEGSIZE];		/* header buffer */
QIC_Header *hptr = (QIC_Header *)hbuff;	/* header structure */
int hsn = -1;				/* segment number of header */
int dhsn = -1;				/* segment number of duplicate header */
int tfd;				/* tape file descriptor */
QIC_Geom geo;				/* tape geometry */
int tvno = 1;				/* tape volume number */
int tvlast;				/* TRUE if last volume in set */
long tvsize = 0;			/* tape volume size in bytes */
long tvtime = 0;			/* tape change time */
char *tvnote = "";			/* tape note */
int doretension = 0;			/* TRUE if we should retension tape */

/* Lookup the badmap for a given track and segment. */
#define BADMAP(t,s)	hptr->qh_badmap[(t)*geo.g_segtrk+(s)]

/* Retrieve values from a character array. */
#define UL_VAL(s,p)	(*((ULONG *)&(s)[p]))
#define US_VAL(s,p)	(*((USHORT *)&(s)[p]))

#define	equal(s1,s2)	(strcmp(s1, s2) == 0)


/*
 *  Print tape usage and then leave.
 */
void
usage(void)
{
  fprintf(stderr, "usage: ft [-f device] [-r] [\"description\"]\n");
  exit(1);
}


/*
 *  Check status of tape drive
 */
int
check_stat(int fd, int wr)
{
  int s;
  int sawit = 0;

  /* get tape status */
  if (ioctl(fd, QIOSTATUS, &s) < 0) {
	warnx("could not get drive status");
	return(1);
  }

  /* wait for the tape drive to become ready */
  while ((s & QS_READY) == 0) {
	if (!sawit) {
		warnx("waiting for drive to become ready");
		sawit = 1;
	}
	sleep(2);
	if (ioctl(fd, QIOSTATUS, &s) < 0) {
		warnx("could not get drive status");
		return(1);
	}
  }

  if ((s & QS_FMTOK) == 0) {
	warnx("tape is not formatted");
	return(2);
  }

  if (wr && (s & QS_RDONLY) != 0) {
	warnx("tape is write protected");
	return(3);
  }

  return(0);
}


/*
 *  Convert time_t value to QIC time value.
 */
ULONG
qtimeval(time_t t)
{
  struct tm *tp;
  ULONG r;

  tp = localtime(&t);
  r = 2678400 * tp->tm_mon +
	86400 *(tp->tm_mday-1) +
	 3600 * tp->tm_hour +
	   60 * tp->tm_min +
	        tp->tm_sec;
  r |= (tp->tm_year - 70) << 25;
  return(r);
}


/*
 *  Return tm struct from QIC date format.
 */
struct tm *
qtime(UCHAR *qt)
{
  ULONG *vp = (ULONG *)qt;
  struct tm t;
  ULONG v;
  time_t tv;

  v = *vp;
  t.tm_year = ((v >> 25) & 0x7f)+70; v &= 0x1ffffff;
  t.tm_mon = v / 2678400; v %= 2678400;
  t.tm_mday = v / 86400 + 1; v %= 86400;
  t.tm_hour = v / 3600; v %= 3600;
  t.tm_min = v / 60; v %= 60;
  t.tm_sec = v;
  t.tm_wday = 0;	/* XXX - let mktime do the real work */
  t.tm_yday = 0;
  t.tm_isdst = 0;
  t.tm_gmtoff = 0;
  t.tm_zone = NULL;
  tv = mktime(&t);
  return(localtime(&tv));
}


/*
 *  Return a string, zero terminated.
 */
char *qstr(char *str, int nchar)
{
  static char tstr[256];
  strncpy(tstr, str, nchar);
  tstr[nchar] = '\0';
  return(tstr);
}


/*
 *  Read header from tape
 */
int
get_header(int fd)
{
  int r, sn;
  QIC_Segment s;
  int gothdr = 0;

  if (ioctl(fd, QIOGEOM, &geo) < 0) {
	warnx("couldn't determine tape geometry");
	return(1);
  }

  /* Get the header and duplicate */
  for (sn = 0; sn < 16; sn++) {
	s.sg_trk = 0;
	s.sg_seg = sn;
	s.sg_badmap = 0;
	s.sg_data = (UCHAR *)&buff[0];
	ioctl(fd, QIOREAD, &s);
	r = check_parity(s.sg_data, 0, s.sg_crcmap);
	if (s.sg_data[0] == 0x55 && s.sg_data[1] == 0xaa &&
	    s.sg_data[2] == 0x55 && s.sg_data[3] == 0xaa) {
		if (hsn >= 0) {
			dhsn = sn;
			if (!r && !gothdr) {
				warnx("using secondary header");
				bcopy(s.sg_data, hbuff, QCV_SEGSIZE);
				gothdr = 1;
			}
			break;
		}
		hsn = sn;
		if (!r) {
			bcopy(s.sg_data, hbuff, QCV_SEGSIZE);
			gothdr = 1;
		} else {
			warnx("too many errors in primary header");
		}
	}
  }

  if (!gothdr) {
	warnx("couldn't read header segment");
	ioctl(fd, QIOREWIND);
	return(1);
  }

  return(0);
}


/*
 *  Open /dev/tty and ask for next volume.
 */
void
ask_vol(int vn)
{
  int fd;
  char c;

  if ((fd = open("/dev/tty", 2)) < 0)
	errx(1, "argh!! can't open /dev/tty");

  fprintf(stderr, "Insert ftfilt volume %02d and press enter:", vn);
  read(fd, &c, 1);
  close(fd);
}


/*
 *  Return the name of the tape only.
 */
void
do_getname(void)
{
  if (check_stat(tfd, 0)) exit(1);
  if (get_header(tfd)) exit(1);
  fprintf(stderr, "\"%s\" - %s",
		qstr(hptr->qh_tname,44), asctime(qtime(hptr->qh_chgdate)));
}


/*
 *  Extract data from tape to stdout.
 */
void
do_read(void)
{
  int sno, vno, sbytes, eccfails;
  long curpos;
  char *hname;
  QIC_Segment s;

  /* Process multiple volumes if necessary */
  vno = 1;
  for (;;) {
	if (check_stat(tfd, 0)) {
		ask_vol(vno);
		continue;
	}

	if (doretension) {
		ioctl(tfd, QIOBOT);
		ioctl(tfd, QIOEOT);
		ioctl(tfd, QIOBOT);
	}

	if (get_header(tfd)) {
		ask_vol(vno);
		continue;
	}

	/* extract volume and header info from label */
	hname = hptr->qh_tname;
	hname[43] = '\0';
	tvno = atoi(&hname[11]);
	tvlast = (hname[10] == '*') ? 1 : 0;
	tvsize = atoi(&hname[14]);
	tvnote = &hname[25];
	if (vno != tvno || strncmp(hname, "ftfilt", 6) != 0) {
		fprintf(stderr, "Incorrect volume inserted.  This tape is:\n");
		fprintf(stderr,"\"%s\" - %s\n", hname,
				asctime(qtime(hptr->qh_chgdate)));
		ask_vol(vno);
		continue;
	}

	/* Process this volume */
	curpos = 0;
	eccfails = 0;
	sno = hptr->qh_first;
	while (tvsize > 0) {
		s.sg_trk = sno / geo.g_segtrk;
		s.sg_seg = sno % geo.g_segtrk;
		s.sg_badmap = BADMAP(s.sg_trk,s.sg_seg);
		sbytes = sect_bytes(s.sg_badmap) - QCV_ECCSIZE;
		s.sg_data = (UCHAR *)&buff[0];
		if (sbytes <= 0) {
			sno++;
			continue;
		}
		if (ioctl(tfd, QIOREAD, &s) < 0) perror("QIOREAD");

		if (check_parity(s.sg_data, s.sg_badmap, s.sg_crcmap)) {
			if (++eccfails <= 5) {
				warnx("retry %d at segment %d byte %ld",
					eccfails, sno, curpos);
				continue;
			} else
				warnx(
				"*** ecc failure in segment %d at byte %ld",
						sno, curpos);
		}
		if (tvsize < sbytes) sbytes = tvsize;
		write(1, s.sg_data, sbytes);
		tvsize -= sbytes;
		curpos += sbytes;
		sno++;
		eccfails = 0;
	}
	if (tvlast) break;
	ioctl(tfd, QIOREWIND);
	ask_vol(++vno);
  }
}


/*
 *  Dump data from stdin to tape.
 */
void
do_write(void)
{
  int sno, vno, amt, sbytes;
  int maxseg, r;
  ULONG qnow;
  QIC_Segment s;
  char tmpstr[80];

  qnow = qtimeval(time(NULL));
  vno = 1;

  for (;;) {
	if (check_stat(tfd, 1)) {
		ask_vol(vno);
		continue;
	}

	if (doretension) {
		ioctl(tfd, QIOBOT);
		ioctl(tfd, QIOEOT);
		ioctl(tfd, QIOBOT);
	}

	if (get_header(tfd)) {
		ask_vol(vno);
		continue;
	}

	maxseg = geo.g_segtrk * geo.g_trktape - 1;
	sno = hptr->qh_first;
	tvno = vno;
	tvsize = 0;
	tvlast = 0;

	/* Process until end of volume or end of data */
	for (sno = hptr->qh_first; sno < maxseg && tvlast == 0; ++sno) {
		/* Prepare to load the next segment */
		s.sg_trk = sno / geo.g_segtrk;
		s.sg_seg = sno % geo.g_segtrk;
		s.sg_badmap = BADMAP(s.sg_trk,s.sg_seg);
		sbytes = sect_bytes(s.sg_badmap) - QCV_ECCSIZE;
		s.sg_data = (UCHAR *)&buff[0];

		/* Ugh.  Loop to get the full amt. */
		for (amt = 0; amt < sbytes; amt += r) {
			r = read(0, &s.sg_data[amt], sbytes - amt);
			if (r <= 0) {
				tvlast = 1;
				break;
			}
		}

		/* skip the segment if *all* sectors are flagged as bad */
		if (amt) {
			if (amt < sbytes)
				bzero(&s.sg_data[amt], sbytes - amt);
			r = set_parity(s.sg_data, s.sg_badmap);
			if (r) warnx("** warning: ecc problem !!");
			if (ioctl(tfd, QIOWRITE, &s) < 0)
				err(1, "QIOWRITE");
			tvsize += amt;
		}
	}

	/* Build new header info */
	/* ftfilt vol*xx yyyyyyyyyy note56789012345678  */
	/* 01234567890123456789012345678901234567890123 */

	sprintf(tmpstr, "ftfilt vol%s%02d %010ld %s",
		(tvlast) ? "*" : " ", tvno, tvsize, tvnote);
	strncpy(hptr->qh_tname, tmpstr, 44);
	UL_VAL(hptr->qh_chgdate,0) = qnow;

	/* Update the header for this volume */
	if (hsn >= 0) {
		s.sg_trk = hsn / geo.g_segtrk;
		s.sg_seg = hsn % geo.g_segtrk;
		s.sg_badmap = 0;
		s.sg_data = (UCHAR *)hbuff;
		r = set_parity(s.sg_data, s.sg_badmap);
		if (r) warnx("** warning: header ecc problem !!");
		if (ioctl(tfd, QIOWRITE, &s) < 0)
			err(1, "QIOWRITE");
	}
	if (dhsn >= 0) {
		s.sg_trk = dhsn / geo.g_segtrk;
		s.sg_seg = dhsn % geo.g_segtrk;
		s.sg_badmap = 0;
		s.sg_data = (UCHAR *)hbuff;
		r = set_parity(s.sg_data, s.sg_badmap);
		if (r) warnx("** warning: duphdr ecc problem !!");
		if (ioctl(tfd, QIOWRITE, &s) < 0)
			err(1, "QIOWRITE");
	}
	ioctl(tfd, QIOREWIND);
	if (tvlast) break;
	ask_vol(++vno);
  }
}


/*
 *  Entry.
 */
int
main(int argc, char *argv[])
{
  int i;
  char *tape, *getenv();


  /* Get device from environment, command line will override. */
  if ((tape = getenv("TAPE")) == NULL) tape = DEFQIC;

  /* Process args. */
  for (i = 1; i < argc; i++) {
	if (argv[i][0] != '-') break;
	switch (argv[i][1]) {
	    case 'f':
	    case 't':
		if (i == (argc - 1)) usage();
		tape = argv[++i];
		break;
	    case 'r':
		doretension = 1;
		break;
	    default:
		usage();
	}
  }
  if (i < (argc - 1)) usage();
  if (i < argc) {
	tvnote = argv[i];
	if (strlen(tvnote) > 18) argv[i][18] = '\0';
  }

  /* Open the tape device */
  if ((tfd = open(tape, 2)) < 0)
	err(1, "%s", tape);

  if (!isatty(0))
	do_write();
  else if (!isatty(1))
	do_read();
  else
	do_getname();

  close(tfd);
  exit(0);
}
